File: hw2/src/base_transport.cpp
#include <string>
#include "base_transport.h"
#include "input_helper.h"
#include "plane.h"
#include "train.h"
#include "ship.h"

BaseTransport::BaseTransport(int speed, float distance)
    : m_speed(speed), m_distance(distance)
{
}

BaseTransport* BaseTransport::Create(std::istream& in)
{
    std::string type;
    if (!(in >> type))
    {
        InputHelper::InvalidInput();
    }

    int speed;
    float distance;
    if (!(in >> speed >> distance))
    {
        InputHelper::InvalidInput();
    }

    if (type == "plane")
    {
        int maxDistance;
        int carryingCapacity;
        if (!(in >> maxDistance >> carryingCapacity))
        {
            InputHelper::InvalidInput();
        }

        return new Plane(speed, distance, maxDistance, carryingCapacity);
    }
    else if (type == "train")
    {
        int carsCount;
        if (!(in >> carsCount))
        {
            InputHelper::InvalidInput();
        }

        return new Train(speed, distance, carsCount);
    }
    else if (type == "ship")
    {
        int displacement;
        std::string vesselTypeStr;
        if (!(in >> displacement >> vesselTypeStr))
        {
            InputHelper::InvalidInput();
        }

        Ship::VesselType vesselType;
        if (vesselTypeStr == "liner")
        {
            vesselType = Ship::VesselType::Liner;
        }
        else if (vesselTypeStr == "tow")
        {
            vesselType = Ship::VesselType::Tow;
        }
        else if (vesselTypeStr == "tanker")
        {
            vesselType = Ship::VesselType::Tanker;
        }
        else
        {
            InputHelper::InvalidInput();
            return nullptr;
        }

        return new Ship(speed, distance, displacement, vesselType);
    }
    else
    {
        InputHelper::InvalidInput();
        return nullptr;
    }
}

BaseTransport* BaseTransport::CreateRnd(std::istream& in, Random& rng)
{
    auto speed = rng.Next(1000);
    auto distance = rng.NextFloat(5000);
    switch (rng.Next(3))
    {
    case 0:
        return new Plane(speed, distance, rng.Next(30000), rng.Next(1000));

    case 1:
        return new Train(speed, distance, rng.Next(40));

    case 2:
        return new Ship(speed, distance, rng.Next(30000), static_cast<Ship::VesselType>(rng.Next(static_cast<int>(Ship::VesselType::MaxIndex))));

    default:
        return nullptr;
    }
}

float BaseTransport::GetIdealTime()
{
    return m_distance / m_speed;
}

void BaseTransport::Out(std::ostream& out)
{
    out << GetName() << ": speed = " << m_speed << ", distance = " << m_distance << ", ideal time = " << GetIdealTime();
}

File: hw3/src/base_transport.py
'''
base_transport.py - содержит описание обобщенного транспорта
'''

import input_helper
import random

# Обобщенный транспорт
class BaseTransport:
    # Инициализация транспорта
    def __init__(self, speed, distance):
        self.speed = speed
        self.distance = distance

    # Идеальное время прохождения пути
    def get_ideal_time(self):
        return self.distance / self.speed

    # Конвертация в строку
    def __str__(self):
        return f'{type(self).__name__}: speed = {self.speed}, distance = {self.distance}, ideal time = {self.get_ideal_time()}'

from plane import *
from train import *
from ship import *
# Создание транспорта из строки
@staticmethod
def create(s):
    splitted = s.split()
    try:
        speed = int(splitted[1])
        distance = float(splitted[2])
        if splitted[0] == 'plane':
            max_distance = int(splitted[3])
            carrying_capacity = int(splitted[4])
            return Plane(speed, distance, max_distance, carrying_capacity)
        elif splitted[0] == 'train':
            cars_count = int(splitted[3])
            return Train(speed, distance, cars_count)
        elif splitted[0] == 'ship':
            displacement = int(splitted[3])
            vessel_type = Ship.VesselType[splitted[4]]
            return Ship(speed, distance, displacement, vessel_type)
        else:
            raise Exception()
    except Exception:
        input_helper.invalid_input()
BaseTransport.create = create
del create

# Создание транспорта (случайно)
@staticmethod
def create_rnd():
    speed = random.randint(1, 1000)
    distance = random.uniform(0, 5000)
    transport_type = random.randint(1, 3)
    if transport_type == 1:
        return Plane(speed, distance, random.randint(0, 30000), random.randint(0, 1000))
    elif transport_type == 2:
        return Train(speed, distance, random.randint(0, 40))
    else:
        return Ship(speed, distance, random.randint(0, 30000), random.choice(list(Ship.VesselType)))
BaseTransport.create_rnd = create_rnd
del create_rnd
